src/server/engine_shard_set.cc

```diff
@@ -56,6 +71,67 @@ EngineShard::Stats& EngineShard::Stats::operator+=(const EngineShard::Stats& o)
   return *this;
 }
 
+// This function checks 3 things:
+// 1. Don't try memory fragmentation if we don't use "enough" memory (control by
+// mem_defrag_threshold flag)
+// 2. That we had change in memory usage - to prevent endless loop - out of scope for now
+// 3. in case the above is OK, make sure that we have a "gap" between usage and commited memory
+// (control by commit_use_threshold flag)
+bool EngineShard::DefragTaskArgs::IsRequired() {
+  static const uint64_t mem_size = max_memory_limit;
+  static const float threshold_mem = mem_size / GetFlag(FLAGS_mem_defrag_threshold);
+  static const float commit_use_threshold = GetFlag(FLAGS_commit_use_threshold);
+
+  if (cursor > 0) {
+    return true;
+  }
+
+  uint64_t commited = GetMallocCurrentCommitted();
+  uint64_t mem_in_use = used_mem_current.load(memory_order_relaxed);
+
+  // we want to make sure that we are not running this to many times - i.e.
+  // if there was no change to the memory, don't run this
+  if (threshold_mem < commited && mem_in_use != 0 &&
+      (float(mem_in_use * commit_use_threshold) < commited)) {
+    // we have way more commited then actual usage
+    return true;
+  }
```

Comment:
GetMallocCurrentCommitted - is mallocs currently commited memory (how much it really stores physically)

mem_in_use - Is what we counted manually to be used? From what I could follow from the code, this is incrementally counted by the mi_memory_resource

`threshold_mem < commited` doesn't match with the description.  This makes sure we've used at least threshold.
> Threshold level to run memory fragmentation between total available memory and currently commited memory


If I understand right from the description, this is how much memory we must have left before running defrag? Like if `threshold < available - commited` we see that, oh no, we have very few memory left, we need to act quickly?

It makes sense how you've implemented it now, as there is no point in running for small usages.

In general, I can't judge what can/would be effective. I guess this just needs to be done with fine tuning once the implementation is ready.

Reply:
This uses 3 parameters:
1. The memory that is available to the application. 
2. The memory that mimalloc committed to (this assumes that it is about the same as RSS memory).
3. The memory we currently using.
When what we are trying to test here is:
- Are we using enough memory from the OS (for example if only 10MB are used, then its really not worth doing something if we have 10GB available). 
- If the above is true, then what is the relation between the committed and used memory, if this ration is big enough, then we need to defrag it so that we will have a better ratio (we want to approach committed = used)